Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/survey/Surveys.kt

```diff
@@ -34,18 +37,37 @@ class Surveys @Inject constructor(
     }
 
     suspend fun requestDetails(type: Type): Survey {
+        val config = appConfigProvider.getAppConfig().survey
+        Timber.v("Requested survey: %s", config)
+        oneTimePasswordRepo.otpAuthorizationResult?.apply {
+            if (authorized && redeemedAt.toDateTime().monthOfYear() == Instant.now().toDateTime().monthOfYear()) {
+                throw SurveyException(SurveyException.Type.ALREADY_PARTICIPATED_THIS_MONTH)
+            }
+        }
+
+        // generate OTP
+        val oneTimePassword = oneTimePasswordRepo.otp ?: oneTimePasswordRepo.generateOTP()
 
-        // TODO adjust for server com
-        // Just to have a glimpse at the loading spinner
-        delay(Seconds.THREE.toStandardDuration().millis)
+        // check device
+        val attestationResult = deviceAttestation.attest(object : DeviceAttestation.Request {
+            override val scenarioPayload: ByteArray
+                get() = oneTimePassword.payloadForRequest
+        })
+        attestationResult.requirePass(config.safetyNetRequirements)
 
-        // TODO: generate and authenticate real otp
-        val otp = UUID.randomUUID()
+        // request validation from server
+        val errorCode = surveyServer.authOTP(oneTimePassword, attestationResult).errorCode
+        val result = OTPAuthorizationResult(oneTimePassword.uuid, errorCode == null, Instant.now())
```

Comment:
(last minor) better use TimeStamper here and above (for Instant.now)

Reply:
fixed it